### DISCUSSION TITLE
swmm execute make corrupt report file

### DISCUSSION THREAD

**Discussion by MohammadTahirGhori** [pyswmm/pyswmm#328] [Q&A]
Hello all

@jennwuu 

I run swmm through sim.execute() then I read report file of swmm for flood depth to calculate dmage cost in an optimization process. I run 10000 times through platypus NSGA II algorithm. 
After the algorithm run all the code and swmm for about 200 times then there is an error that report file cannot be open. When I check report file it seems that report file is not complete. When I check the lines before error it seems that when swmm run then the next step is reading report file and report file is not complete. There is something with swmm execute. 
I use pyswmm latest version. Do you think it is related to pyswmm version and how can I solve this problem. 

here is the error lines 

  File "C:\Users\mgh004\Anaconda3\lib\site-packages\platypus\core.py", line 191, in evaluate
    objs = self.function(solution.variables)

  File "D:\Thesis\MSc_Yang\strategy1\scenario1.py", line 540, in cost_benefit
    objectives = Damage_withNBS()

  File "D:\Thesis\MSc_Yang\strategy1\scenario1.py", line 452, in Damage_withNBS
    sim.execute()

  File "C:\Users\mgh004\Anaconda3\lib\site-packages\pyswmm\simulation.py", line 422, in execute
    self._model.swmmExec()

  File "C:\Users\mgh004\Anaconda3\lib\site-packages\pyswmm\swmm5.py", line 156, in swmmExec
    self.swmm_run()

  File "C:\Users\mgh004\Anaconda3\lib\site-packages\pyswmm\swmm5.py", line 179, in swmm_run
    solver.swmm_run(inpfile, rptfile, binfile)

  File "C:\Users\mgh004\Anaconda3\lib\site-packages\swmm\toolkit\solver.py", line 493, in swmm_run
    return _solver.swmm_run(f1, f2, f3)

Exception: 
  ERROR 305: cannot open report file.

**Comment by jennwuu** [comment]
Are you running models individually or are you running models in parallel? The error is indicating invalid report file / report file does not exist. Check your model report file path and confirm that it exists.

**Reply by MohammadTahirGhori** [reply]
Hi 
Thanks for your reply. 
Actually it works for around 200 runs but in between this error arise. I run for 10000 times by NSGA II algorithm. 
During the run bellow arround 200 runs I check the report file is updating respectively by run but after some runs it stopes and rise the error. 

The code is:

    # Run SWMM.
    run_SWMM("D:/Thesis/MSc_Yang/strategy1/Basefile.inp") # baseline simulation should be run only once as multiple runs will give the same result. 
    # Read SWMM simulation result file for flood depths. 
    baseline = node_flood_depth("D:\Thesis\MSc_Yang\strategy1\BaseFile.rpt", [483, 520]) # baseline is a dataframe with two columns - nodeID & max_depth

Then I calculate damage using some functions. 

Then I run swmm with implemented Nature Based Solutions measures 
This part will run 10000 times the above baseline run is once 

    # STEP 6: Run SWMM
    run_SWMM("D:/Thesis/MSc_Yang/strategy1/Basefile_updated.inp")

    # STEP 7: Read SWMM simulation result file for flood depths.
    withNBS = node_flood_depth("D:\Thesis\MSc_Yang\strategy1\Basefile_updated.rpt", [531, 568]) # withNBS is a dataframe with two columns - nodeID & max_depth

Then here is the algorithm which calls for run of swmm 

problem.directions[0]=Problem.MINIMIZE # minimize total cost
problem.directions[1]=Problem.MAXIMIZE # maximize the difference between beseline damage and damage with NBS

problem.function = cost_benefit
algorithm = NSGAII(problem,100) #default the population is 100
algorithm.run(1000) # run the algorithm for 10,000 function evaluations

**Reply by jennwuu** [reply]
Instead of using the same filename for all 10000 model runs, can you use unique file name and try again? e.g. Baseline_0001.rpt

**Reply by MohammadTahirGhori** [reply]
@jennwuu 
Hi 
I tried another method to run swmm and get max depth and pollutants with pyswmm 1.0.1.
I used 
    with Simulation(file_path) as sim:
        for step in sim:
            pass
        J1 = Links(sim)['cs1']
        J1.total_loading

This method I get the max depth and pollutants for all runs. 
Is this the correct way?
I checked the results with manually running swmm and comparing the max depth in report file it was the same.

**Reply by jennwuu** [reply]
@MohammadTahirGhori Can you reshare your code snippet with indents?

**Reply by MohammadTahirGhori** [reply]
@jennwuu 
Hi
Sorry, I didnt get your question. 
I had two ways to run optimization and get max depth at nodes.
1- Sim.execute and read from the reported file 
 def run_SWMM(SWMM_input):
     sim = Simulation(SWMM_input)
     sim.execute()

2- sim.pass and ask for max depth directly 
def run_SWMM(SWMM_input):
    with Simulation(SWMM_input) as sim:
        for step in sim:
            pass
The first one had an error after some runs but the second method is working for me now for any number of runs.

**Comment by MohammadTahirGhori** [comment]
@bemcdonnell 
Can you please comment in this method. This is another method that I use to directly get the max depth from specific lines of report file and it is fast than calling max depth using Simulation and advance step method which takes too much time. 
But here it raise an error after around 170 run of swmm.  For approx 170 runs it is ok and creates report file and read it but after that it says that cannot open report file and report file is also not complete when I check it.
